file/reed_solomon_file.go

```diff
@@ -0,0 +1,466 @@
+	}
+
+	b := rsDagInstance.buff.Bytes()
+	offset := fNode.StartOffset
+	if rsDagInstance.offset != offset {
+		return nil, errors.New("offset from the given FileNode is invalid.")
+	}
+	newOffset := offset + uint64(fNode.NodeSize())
+	if newOffset > uint64(len(b)) {
+		return nil, errors.New("new offset is greater than buffer size.")
+	}
+	dr := bytes.NewReader(b[offset:newOffset])
+
+	rsDagInstance.offset = newOffset
+	return &rsFile{
+		DagReader: &uio.ReedSolomonDagReader{dr},
+	}, nil
+}
+
+func NewReedSolomonStandaloneFile(ctx context.Context, root ipld.Node, dataNode ipld.Node, dserv ipld.DAGService, opts UnixfsFileOptions,
+	metaStruct *MetadataStruct) (files.Node, error) {
+	rsMeta := metaStruct.RsMeta
+	dr, mrs, _, err := uio.NewReedSolomonDagReader(ctx, dataNode, dserv, rsMeta.NumData, rsMeta.NumParity,
+		rsMeta.FileSize, rsMeta.IsDir, opts.RepairShards)
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkAndRecoverShards(ctx, root, dserv, opts, mrs)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rsFile{
+		DagReader: dr,
+	}, nil
+}
+
+func checkAndRecoverShards(ctx context.Context, root ipld.Node, dserv ipld.DAGService, opts UnixfsFileOptions, mrs []io.Reader) error {
+	// Check which ones need recovery
+	var recovered []io.Reader
+	var rcids []cid.Cid
+	for i, mr := range mrs {
+		if mr != nil {
+			recovered = append(recovered, mr)
+			rcids = append(rcids, opts.RepairShards[i])
+		}
+	}
+	if len(recovered) > 0 {
+		err := addRecoveredShards(ctx, root, dserv, recovered, rcids)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addRecoveredShards mimics adding reed solomon shards anew according to the
+// original adder options.rootNode ipld.Node
+func addRecoveredShards(ctx context.Context, rootNode ipld.Node, ds ipld.DAGService,
+	recovered []io.Reader, rcids []cid.Cid) error {
+	b, err := uio.GetMetaDataFromDagRoot(ctx, rootNode, ds)
+	if err != nil {
+		return err
+	}
+
+	bMeta := util.GetMetadataElement(b)
+	sm, err := ihelper.GetOrDefaultSuperMeta(bMeta)
+	if err != nil {
+		return err
+	}
+
+	// Recover shard has a default size-based chunker
+	sc := fmt.Sprintf("size-%d", sm.ChunkSize)
+	for i, r := range recovered {
+		chnk, err := chunker.FromString(r, sc)
+		if err != nil {
+			return err
+		}
+
+		// Re-create (as much as possible) the original params
+		params := ihelper.DagBuilderParams{
+			Dagserv:   ds,
+			Maxlinks:  int(sm.MaxLinks),
+			ChunkSize: sm.ChunkSize,
+		}
+
+		db, err := params.New(chnk)
+		if err != nil {
+			return err
+		}
+
+		var sn ipld.Node // new shard root node
+		if sm.TrickleFormat {
+			sn, err = trickle.Layout(db)
+		} else {
+			sn, err = balanced.Layout(db)
+		}
+		if err != nil {
+			return err
+		}
+
+		if !rcids[i].Equals(sn.Cid()) {
+			return fmt.Errorf("recovered node [%s] does not match original [%s]",
+				sn.Cid().String(), rcids[i].String())
+		}
+	}
+
+	return nil
+}
+
+var _ files.Directory = &rsDirectory{}
+var _ files.File = &rsFile{}
```

Comment:
what do these do? just to make sure they are initialized?

Reply:
deleted